In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')
%matplotlib inline

pd.set_option('display.max_columns', 200)
pd.set_option('display.max_rows', 200)
pd.set_option('display.width', 1000)

In [2]:
df = pd.read_csv("AB_NYC_2019.csv")

In [3]:
df.head()

id                                              name  host_id    host_name neighbourhood_group neighbourhood  latitude  longitude        room_type  price  minimum_nights  number_of_reviews last_review  reviews_per_month  calculated_host_listings_count  availability_365
0  2539                Clean & quiet apt home by the park     2787         John            Brooklyn    Kensington  40.64749  -73.97237     Private room    149               1                  9  19-10-2018               0.21                               6               365
1  2595                             Skylit Midtown Castle     2845     Jennifer           Manhattan       Midtown  40.75362  -73.98377  Entire home/apt    225               1                 45  21-05-2019               0.38                               2               355
2  3647               THE VILLAGE OF HARLEM....NEW YORK !     4632    Elisabeth           Manhattan        Harlem  40.80902  -73.94190     Private room    150               3                  0         NaN                NaN                               1               365
3  3831                   Cozy Entire Floor of Brownstone     4869  LisaRoxanne            Brooklyn  Clinton Hill  40.68514  -73.95976  Entire home/apt     89               1                270  05-07-2019               4.64                               1               194
4  5022  Entire Apt: Spacious Studio/Loft by central park     7192        Laura           Manhattan   East Harlem  40.79851  -73.94399  Entire home/apt     80              10                  9  19-11-2018               0.10                               1                 0

In [4]:
df.shape

(48895, 16)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48895 entries, 0 to 48894
Data columns (total 16 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   id                              48895 non-null  int64  
 1   name                            48879 non-null  object 
 2   host_id                         48895 non-null  int64  
 3   host_name                       48874 non-null  object 
 4   neighbourhood_group             48895 non-null  object 
 5   neighbourhood                   48895 non-null  object 
 6   latitude                        48895 non-null  float64
 7   longitude                       48895 non-null  float64
 8   room_type                       48895 non-null  object 
 9   price                           48895 non-null  int64  
 10  minimum_nights                  48895 non-null  int64  
 11  number_of_reviews               48895 non-null  int64  
 12  last_review                     

In [6]:
df.describe()

id       host_id      latitude     longitude         price  minimum_nights  number_of_reviews  reviews_per_month  calculated_host_listings_count  availability_365
count  4.889500e+04  4.889500e+04  48895.000000  48895.000000  48895.000000    48895.000000       48895.000000       38843.000000                    48895.000000      48895.000000
mean   1.901714e+07  6.762001e+07     40.728949    -73.952170    152.720687        7.029962          23.274466           1.373221                        7.143982        112.781327
std    1.098311e+07  7.861097e+07      0.054530      0.046157    240.154170       20.510550          44.550582           1.680442                       32.952519        131.622289
min    2.539000e+03  2.438000e+03     40.499790    -74.244420      0.000000        1.000000           0.000000           0.010000                        1.000000          0.000000
25%    9.471945e+06  7.822033e+06     40.690100    -73.983070     69.000000        1.000000           1.000000           0.190000                        1.000000          0.000000
50%    1.967728e+07  3.079382e+07     40.723070    -73.955680    106.000000        3.000000           5.000000           0.720000                        1.000000         45.000000
75%    2.915218e+07  1.074344e+08     40.763115    -73.936275    175.000000        5.000000          24.000000           2.020000                        2.000000        227.000000
max    3.648724e+07  2.743213e+08     40.913060    -73.712990  10000.000000     1250.000000         629.000000          58.500000                      327.000000        365.000000

In [7]:
def describe_with_range(df):
    des = df.describe()
    des.loc["range"] = des.loc['max'] - des.loc['min']
    return des

# print (describe_new())

In [8]:
describe_with_range(df)

id       host_id      latitude     longitude         price  minimum_nights  number_of_reviews  reviews_per_month  calculated_host_listings_count  availability_365
count  4.889500e+04  4.889500e+04  48895.000000  48895.000000  48895.000000    48895.000000       48895.000000       38843.000000                    48895.000000      48895.000000
mean   1.901714e+07  6.762001e+07     40.728949    -73.952170    152.720687        7.029962          23.274466           1.373221                        7.143982        112.781327
std    1.098311e+07  7.861097e+07      0.054530      0.046157    240.154170       20.510550          44.550582           1.680442                       32.952519        131.622289
min    2.539000e+03  2.438000e+03     40.499790    -74.244420      0.000000        1.000000           0.000000           0.010000                        1.000000          0.000000
25%    9.471945e+06  7.822033e+06     40.690100    -73.983070     69.000000        1.000000           1.000000           0.190000                        1.000000          0.000000
50%    1.967728e+07  3.079382e+07     40.723070    -73.955680    106.000000        3.000000           5.000000           0.720000                        1.000000         45.000000
75%    2.915218e+07  1.074344e+08     40.763115    -73.936275    175.000000        5.000000          24.000000           2.020000                        2.000000        227.000000
max    3.648724e+07  2.743213e+08     40.913060    -73.712990  10000.000000     1250.000000         629.000000          58.500000                      327.000000        365.000000
range  3.648471e+07  2.743189e+08      0.413270      0.531430  10000.000000     1249.000000         629.000000          58.490000                      326.000000        365.000000

## Empty Values

In [9]:
# Percentage of missing values
round(df.isnull().mean() * 100, 2).sort_values(ascending = False)

last_review                       20.56
reviews_per_month                 20.56
host_name                          0.04
name                               0.03
id                                 0.00
host_id                            0.00
neighbourhood_group                0.00
neighbourhood                      0.00
latitude                           0.00
longitude                          0.00
room_type                          0.00
price                              0.00
minimum_nights                     0.00
number_of_reviews                  0.00
calculated_host_listings_count     0.00
availability_365                   0.00
dtype: float64

In [11]:
1 - df[df.host_name.isnull() | df.name.isnull()].shape[0]/len(df)

0.9992432764086308

We are retaining 99.92 % of data after excluding the rows with null values w.r.t. df.host_name and df.name

In [12]:
df = df[-(df.host_name.isnull() | df.name.isnull())]

In [15]:
df.last_review.isnull() == df.reviews_per_month.isnull()

0        True
1        True
2        True
3        True
4        True
         ... 
48890    True
48891    True
48892    True
48893    True
48894    True
Length: 48858, dtype: bool

In [17]:
len(df.last_review.isnull())

48858

We can confirm that both column df.last_review and df.reviews_per_month are having null values for identical rows

In [21]:
df["last_review"]

0        19-10-2018
1        21-05-2019
2               NaN
3        05-07-2019
4        19-11-2018
            ...    
48890           NaN
48891           NaN
48892           NaN
48893           NaN
48894           NaN
Name: last_review, Length: 48858, dtype: object

df.last_review column is data column but it is interpreted as object dtype. lets change it

In [22]:
df["last_review"] = df["last_review"].astype('datetime64[ns]')

In [23]:
df["last_review"]

0       2018-10-19
1       2019-05-21
2              NaT
3       2019-05-07
4       2018-11-19
           ...    
48890          NaT
48891          NaT
48892          NaT
48893          NaT
48894          NaT
Name: last_review, Length: 48858, dtype: datetime64[ns]

In [27]:
df[df["reviews_per_month"].isnull()]

id                                               name   host_id        host_name neighbourhood_group       neighbourhood  latitude  longitude        room_type  price  minimum_nights  number_of_reviews last_review  reviews_per_month  calculated_host_listings_count  availability_365
2          3647                THE VILLAGE OF HARLEM....NEW YORK !      4632        Elisabeth           Manhattan              Harlem  40.80902  -73.94190     Private room    150               3                  0         NaT                NaN                               1               365
19         7750                  Huge 2 BR Upper East  Cental Park     17985             Sing           Manhattan         East Harlem  40.79685  -73.94872  Entire home/apt    190               7                  0         NaT                NaN                               2               249
26         8700  Magnifique Suite au N de Manhattan - vue Cloitres     26394  Claude & Sophie           Manhattan              Inwood  40.86754  -73.92639     Private room     80               4                  0         NaT                NaN                               1                 0
36        11452                        Clean and Quiet in Brooklyn      7355               Vt            Brooklyn  Bedford-Stuyvesant  40.68876  -73.94312     Private room     35              60                  0         NaT                NaN                               1               365
38        11943                          Country space in the city     45445          Harriet            Brooklyn            Flatbush  40.63702  -73.96327     Private room    150               1                  0         NaT                NaN                               1               365
...         ...                                                ...       ...              ...                 ...                 ...       ...        ...              ...    ...             ...                ...         ...                ...                             ...               ...
48890  36484665    Charming one bedroom - newly renovated rowhouse   8232441          Sabrina            Brooklyn  Bedford-Stuyvesant  40.67853  -73.94995     Private room     70               2                  0         NaT                NaN                               2                 9
48891  36485057      Affordable room in Bushwick/East Williamsburg   6570630          Marisol            Brooklyn            Bushwick  40.70184  -73.93317     Private room     40               4                  0         NaT                NaN                               2                36
48892  36485431            Sunny Studio at Historical Neighborhood  23492952    Ilgar & Aysel           Manhattan              Harlem  40.81475  -73.94867  Entire home/apt    115              10                  0         NaT                NaN                               1                27
48893  36485609               43rd St. Time Square-cozy single bed  30985759              Taz           Manhattan      Hell's Kitchen  40.75751  -73.99112      Shared room     55               1                  0         NaT                NaN                               6                 2
48894  36487245  Trendy duplex in the very heart of Hell's Kitchen  68119814       Christophe           Manhattan      Hell's Kitchen  40.76404  -73.98933     Private room     90               7                  0         NaT                NaN                               1                23

[10037 rows x 16 columns]

As we confirmed that the null rows are same for df.last_review and df.reviews_per_month columns. We can say that last_review is null cause there was no review recorded in the first place

### Handling Missing values

In [28]:
df.reviews_per_month.fillna(0, inplace=True)

In [29]:
round(df.isnull().mean() * 100, 2).sort_values(ascending = False)

last_review                       20.54
id                                 0.00
name                               0.00
host_id                            0.00
host_name                          0.00
neighbourhood_group                0.00
neighbourhood                      0.00
latitude                           0.00
longitude                          0.00
room_type                          0.00
price                              0.00
minimum_nights                     0.00
number_of_reviews                  0.00
reviews_per_month                  0.00
calculated_host_listings_count     0.00
availability_365                   0.00
dtype: float64